docs/sql-ref-datatypes.md

```diff
@@ -314,3 +314,128 @@ SELECT COUNT(*), c2 FROM test GROUP BY c2;
 |        3| Infinity|
 +---------+---------+
 ```
+
+### Type conversion
+
+In general, an expression can contain different data types, type conversion is the transformation of some data types into others in order to solve the expressions. 
+Spark supports both implicit conversions by type coercion and explicit conversions by explicit casting and store assignment casting.
+
+#### Type coercion in operations between different types 
+
+The following matrix shows the resulting type to which they are implicitly converted to resolve an expression involving different data types 
+
+Numeric expresions:
+
+|               |ByteType   |ShortType  |IntegerType |LongType  |FloatType |DoubleType |StringType |
+|---------------|-----------|-----------|------------|----------|----------|-----------|-----------|
+|**ByteType**   |--         |ShortType  |IntegerType |LongType  |FloatType |DoubleType |DoubleType |
+|**ShortType**  |ShortType  |--         |IntegerType |LongType  |FloatType |DoubleType |DoubleType |
+|**IntegerType**|IntegerType|IntegerType|--          |LongType  |FloatType |DoubleType |DoubleType |
+|**LongType**   |LongType   |LongType   |LongType    |--        |FloatType |DoubleType |DoubleType |
+|**FloatType**  |FloatType  |FloatType  |FloatType   |FloatType |--        |DoubleType |DoubleType |
+|**DoubleType** |DoubleType |DoubleType |DoubleType  |DoubleType|DoubleType|--         |DoubleType |
+|**StringType** |DoubleType |DoubleType |DoubleType  |DoubleType|DoubleType|DoubleType |--         |
```

Comment:
This matrix means specific type coercion rules in binary arithmetic operations, right? Cold you add a matrix for type coercion in more general cases, too, like `Table 2-10 Implicit Type Conversion Matrix` in the Oracle doc? You could refer to the rules in `TypeCoercion.scala` https://github.com/apache/spark/blob/master/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TypeCoercion.scala#L983-L1064